Source/JavaScriptCore/inspector/InjectedScript.cpp

```diff
@@ -90,18 +90,16 @@ void InjectedScript::awaitPromise(const String& promiseObjectId, bool returnByVa
     makeAsyncCall(function, WTFMove(callback));
 }
 
-void InjectedScript::callFunctionOn(Protocol::ErrorString& errorString, const String& objectId, const String& expression, const String& arguments, bool returnByValue, bool generatePreview, RefPtr<Protocol::Runtime::RemoteObject>& result, std::optional<bool>& wasThrown)
+void InjectedScript::callFunctionOn(const String& objectId, const String& expression, const String& arguments, bool returnByValue, bool generatePreview, bool awaitPromise, AsyncCallCallback&& callback)
 {
     ScriptFunctionCall function(globalObject(), injectedScriptObject(), "callFunctionOn"_s, inspectorEnvironment()->functionCallHandler());
     function.appendArgument(objectId);
     function.appendArgument(expression);
     function.appendArgument(arguments);
     function.appendArgument(returnByValue);
     function.appendArgument(generatePreview);
-
-    std::optional<int> savedResultIndex;
-    makeEvalCall(errorString, function, result, wasThrown, savedResultIndex);
-    ASSERT(!savedResultIndex);
+    function.appendArgument(awaitPromise);
+    makeAsyncCall(function, WTFMove(callback));
```

Comment:
im a bit concerned with switching from `makeEvalCall` to `makeAsyncCall`

are we 100% sure that the behavior in the non-async (i.e. `awaitPromise: false`) case is the same?

Reply:
Looking at the code, the non-waiting logic should be pretty much the same as before. The tests pass and manual testing doesn't show any regressions. I don't know how else we can validate that.

Reply:
i dont think a lot of the places we currently use `Runtime.callFunctionOn` are covered by tests :/

how much have you tried using Web Inspector with these changes?  AFAICT it seems like all of these user flows could be affected by this change:
- replaying 2D `<canvas>`
- "climbing" an object's prototype chain in the Console
- previewing `get` member functions of an object in the Console
- getting JavaScript completions in the Console
- displaying custom Audit result objects
- toggling CSS classes in the Styles details sidebar panel in the Elements Tab
- scrolling nodes into view via the Elements Tab
- etc.

Reply:
OK, tried all these scenarios, didn't notice any regressions so far.